### PR TITLE
Add variable substitution in ApplicationManifestUtils

### PR DESCRIPTION
**Pull request by luschmar**
Add functionality [Variables to a Manifest](https://docs.cloudfoundry.org/devguide/deploy-apps/manifest-attributes.html#variable-substitution) in ApplicationManifestUtils

### CODE REVIEW

File: cloudfoundry-operations/src/test/java/org/cloudfoundry/operations/applications/ApplicationManifestUtilsTest.java [hunk 1]
```diff
@@ -455,6 +455,22 @@ public void readSingleBuildpack() throws IOException {
         assertThat(actual).isEqualTo(expected);
     }
 
+    @Test
```
Comment (dmikusa-pivotal) [inline cloudfoundry-operations/src/test/java/org/cloudfoundry/operations/applications/ApplicationManifestUtilsTest.java:458]: I'd like to see some more test coverage here. 

Off the top of my head:

1. What happens if the manifest.yml has a placeholder that's not given a value? I feel like this should surface some sort of error.
2. What if there are values provided but no placeholder for them?  This seems ignorable.
3. Are there things one could do to accidentally or intentionally break this functionality? I do see that you're quoting the key as it's put into a regex, it would probably be good to have the test do something that would break w/out quoting.

**Review by dmikusa-pivotal — CHANGES_REQUESTED** [review]

File: cloudfoundry-operations/src/test/java/org/cloudfoundry/operations/applications/ApplicationManifestUtilsTest.java (hunk 1 above)
Comment (luschmar) [reply cloudfoundry-operations/src/test/java/org/cloudfoundry/operations/applications/ApplicationManifestUtilsTest.java:459]: The whole behavior is not specified. My knowledge relies only on this passage [Add Variables to a Manifest](https://docs.cloudfoundry.org/devguide/deploy-apps/manifest-attributes.html#variable-substitution)

1. I do not think; not substituted variables should throw an error. It's no illegal to have an app name like ((abc))
2. I ignore this; 
3. As you can see in some additional tests, there are a lot of edge cases. Since there is no grammar or more specification I cannot preprocess the variable keys or even values.

I know a lot of things can go wrong with this code, when someone tries to rape this functionality. But lets be honest, average users of this substitution will create var-files for every environment with simple text/integer values and use them instead creating multiple manifest for every environment.

My intention for this MR: I have some projects they rely on these substitution stuff used with cf-cli. I want to replace my long bash-script-bloat (blue/green deployment; rolling restart; org cleanup) uses cf-cli and I want some stable cf-java-client replacement. 

Before I rewrite my MR; perhaps you can provide me more insights of this functionality.

Some resources I consulted:
 * https://cli.cloudfoundry.org/en-US/v8/
 * https://v3-apidocs.cloudfoundry.org/version/3.117.0/index.html

Comment (luschmar) [reply cloudfoundry-operations/src/test/java/org/cloudfoundry/operations/applications/ApplicationManifestUtilsTest.java:459]: I changed the approach. I just search for variables and replace them accordingly. And added more tests

File: cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/ApplicationManifestUtils.java [hunk 2]
```diff
@@ -125,21 +149,32 @@ public static void write(OutputStream out, List<ApplicationManifest> application
         }
     }
 
-    private static <T> void as(Map<String, Object> payload, String key, Function<Object, T> mapper, Consumer<T> consumer) {
+    private static <T> void as(Map<String, Object> payload, String key, Map<String, String> variables, Function<Object, T> mapper, Consumer<T> consumer) {
         Optional.ofNullable(payload.get(key))
+            .map(o -> {
+                if(o instanceof String) {
+                    Matcher m = FIND_VARIABLE_REGEX.matcher(((String) o));
```
Comment (dmikusa-pivotal) [inline cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/applications/ApplicationManifestUtils.java:156]:
```suggestion
                    Matcher m = FIND_VARIABLE_REGEX.matcher((String) o);
```

I don't believe you need the extra `(..)`.

**Review by dmikusa-pivotal — APPROVED** [review]

**Comment by dmikusa-pivotal** [comment]
Thanks @luschmar !
